### ISSUE TITLE
Badge Request: GitHub Container Registry

### ISSUE THREAD

**Issue by cicirello** [badges/shields#5594]
:clipboard: **Description**

GitHub recently introduced the GitHub Container Registry (see: https://github.blog/2020-09-01-introducing-github-container-registry/).  It would be nice if there were badges available similar to those that shields has for Docker image version, image size, and pulls, but for the new GitHub Container Registry.

:link: **Data**

I'm not sure if GitHub has an API for this yet.

:microphone: **Motivation**

Motivation is the same as that of the equivalent badges that already exist for Docker Hub, but targeting the GitHub Container Registry instead.  Providing a nice concise visual summary of some of the key statistics one might be interested in with a Docker image, but for images available in the GCR.

**Comment by chris48s** [comment]
Given this is essentially a feature related to the package registry, I suspect this one is probably going to end up blocked on #4169 (tldr: the github tokens we have don't have the necessary scope to read package metadata) once the API endpoints are available, but we'll see..

**Comment by adriens** [comment]
🙏

**Comment by eggplants** [comment]
To get the pull counter of the image published on ghcr, you can use [PackageStatistics](https://docs.github.com/graphql/reference/objects#packagestatistics) or [PackageVersionStatistics](https://docs.github.com/graphql/reference/objects#packageversionstatistics) in GraphQL API (needs auth).

To get information of tags and sizes of public images on ghcr.io, you can use `https://ghcr.io/v2/` as [Docker Registry HTTP API V2](https://docs.docker.com/registry/spec/api/)'s endpoint like DockerHub.

Here is my example with curl:

```bash
# public image's {USER}/{IMAGE}
USER_IMAGE=eggplants/asciiquarium-docker

# get fake token ('{"token":"***"}' -> '***')
TOKEN="$(
  curl "https://ghcr.io/token?scope=repository:${USER_IMAGE}:pull" |
  awk -F'"' '$0=$4'
)"

_curl(){ curl -H "Authorization: Bearer ${TOKEN}" "$1" }

# get tags
_curl "https://ghcr.io/v2/${USER_IMAGE}/tags/list"
# >>> {"name":"eggplants/asciiquarium-docker","tags":["0.0.0","0.0","latest"]}

# get manifest of the latest image
_curl "https://ghcr.io/v2/${USER_IMAGE}/manifests/latest"
# >>>
# {
#    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
#    "schemaVersion": 2,
#    "config": {
#       "mediaType": "application/vnd.docker.container.image.v1+json",
#       "digest": "sha256:fb4552eacdc8e01082f87bae6a0fec6e1b61630f0a1e5a0e6a699d05ff0337fd",
#       "size": 5551
#    },
#    "layers": [
#       {
#          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
#          "digest": "sha256:4be315f6562fccf08fd6c749557e31e45ab6d987370e20e2c4933ddb04ddd5ff",
#          "size": 27140664
#       }, ...
```

**Comment by calebcartwright** [comment]
> The number of pulls seems difficult to fetch, but it would be relatively easy to get just the tags and their image sizes, which do not require authentication.

Gotcha, thanks for the clarification! In that case think it would probably make sense at this point to extract a separate issue for version and size badges to discuss the viability of the proposed approach, and we narrow the scope of this blocked one to pulls

**Comment by eggplants** [comment]
I have implemented [eggplants/ghcr-badge](https://github.com/eggplants/ghcr-badge) for my experiment:
![tags](https://ghcr-badge.herokuapp.com/eggplants/ghcr-badge/tags?color=#44cc11)
![latest_tag](https://ghcr-badge.herokuapp.com/eggplants/ghcr-badge/latest_tag?color=#44cc11)
![size](https://ghcr-badge.herokuapp.com/eggplants/ghcr-badge/size?tag=latest&color=#44cc11)

**Comment by asssaf** [comment]
you can also hard code the token per image:
```
curl -H "Authorization: Bearer $(echo "v1:$USER_IMAGE:0" | base64)" "https://ghcr.io/v2/$USER_IMAGE/tags/list"
```

**Comment by eggplants** [comment]
@asssaf Thanks! https://github.com/eggplants/ghcr-badge/commit/f5504c4bbe03fbc8e0899fb8c2ba917f5b68c06e

**Comment by simonmicro** [comment]
With the coming shutdown of free organizations on Docker Hub, we also plan to migrate to ghcr.io - what is the state of this implementation right now? I really do not want to stress this :grin:  ! Take your time :heart_on_fire:

**Comment by trajan0x** [comment]
+1 on this

**Comment by nvuillam** [comment]
i'm trying with GraphQL api, and even with a Personal Access Token set, I don't receive any package in the results
Would anyone have a tip ? :)

```graphql
query {
  organization(login: "oxsecurity") {
      name
      isVerified
    packages(first: 100) {
        nodes {
            name
            packageType
            statistics {
                downloadsTotalCount
            }
        }
    }
  }
}
```

```json
{
    "data": {
        "organization": {
            "name": "OX Security",
            "isVerified": true,
            "packages": {
                "nodes": []
            }
        }
    }
}
```

**Comment by eggplants** [comment]
@nvuillam It doesn't seem to work.
https://github.com/eggplants/ghcr-badge/issues/72#issuecomment-1665988472

**Comment by nvuillam** [comment]
> @nvuillam It doesn't seem to work. [eggplants/ghcr-badge#72 (comment)](https://github.com/eggplants/ghcr-badge/issues/72#issuecomment-1665988472)

That' really sad :(

Thanks for the reply !

**Comment by cicirello** [comment]
> > @nvuillam It doesn't seem to work. [eggplants/ghcr-badge#72 (comment)](https://github.com/eggplants/ghcr-badge/issues/72#issuecomment-1665988472)
> 
> That' really sad :(
> 
> Thanks for the reply !

@nvuillam try a repository query (instead of an organization query) to get the packages for the megalinter repository. It looks like all of your public packages are associated with that repository, so that might work. I didn't try this so I don't know for sure if it will work, but I think it might.

**Comment by eggplants** [comment]
@cicirello You can try this query from <https://docs.github.com/en/graphql/overview/explorer>. And `repository.packages` is also empty.
```gql
query {
  repository(owner: "oxsecurity", name: "megalinter") {
    name    
    packages(first: 1) {
        nodes {
            name
            packageType
            statistics {
                downloadsTotalCount
            }
        }
    }
  }
}
```

![image](https://github.com/badges/shields/assets/42153744/c70a6646-7da9-4463-a7c5-b2a41606a861)

Note that `oxsecurity/megalinter` has [165 Packages](https://github.com/orgs/oxsecurity/packages?repo_name=megalinter).

**Comment by nvuillam** [comment]
I confirm, I still did not find any way to query stats for MegaLinter packages :/

**Comment by jovezhong** [comment]
It's sad that the GraphQL API from GitHub doesn't work.

I ended up with a [tool](https://www.crul.com/) to parse the HTML page and list only named versions
```
range 1 5 page
|| open https://github.com/timeplus-io/proton/pkgs/container/proton/versions?page=$page$ --cache false --html --hashtml|| filter "(nodeName=='LI') and attributes.class =='Box-row'" || sequence
|| html innerHTML || filter "(_html.nodeName == 'A')" ||rename _html.innerText version || startswith version '1.3.'  || html innerHTML || filter "(_html.nodeName == 'SPAN' and _html.attributes.style =='white-space:nowrap;')"||rename _html.innerText downloads || replace downloads 'Version downloads' ''||table version downloads

```
<img width="1098" alt="image" src="https://github.com/badges/shields/assets/5076438/90892584-e404-4d49-982b-9c5b934b2154">

**Comment by nvuillam** [comment]
I thought to build something similar, like [github-dependent-infos](https://github.com/nvuillam/github-dependents-info) scraps to get the number and list of dependent repos...

But a working API would be the best, we're just looking for simple numbers :(

**Comment by TheCaptain989** [comment]
I _would_ love support for tags.  My main use case is here: https://github.com/linuxserver/docker-mods/pkgs/container/mods/
To pick one tag: https://github.com/linuxserver/docker-mods/pkgs/container/mods/227277266?tag=universal-docker

**Comment by TheCaptain989** [comment]
> I made a JSON endpoint for the total downloads that can be used to make a badge. Submit an issue if you want me to also track counts for each version/tag.
> 
> [ghcr-pulls](https://github.com/ipitio/ghcr-pulls)

I _would_ love support for tags.  My main use case is here: https://github.com/linuxserver/docker-mods/pkgs/container/mods/
To pick one tag: https://github.com/linuxserver/docker-mods/pkgs/container/mods/227277266?tag=universal-docker

Trying to mess with it now...

**Comment by ipitio** [comment]
@TheCaptain989 I've got version support mostly implemented!

I'm also expanding support to all package types on GitHub Packages and want to add sizes for each version to the json, but I haven't looked into whether it's possible to get the sizes via rest api/scraping/graphql.

Does anyone how/if I could get the sizes? ghcr-badge and, at the moment, my repo cheat for `container`s by using their docker manifests. Cheating on the other ones would also be fine!

**Comment by eggplants** [comment]
@ipitio I read [your current approach](https://github.com/ipitio/ghcr-pulls/blob/b8b57302fdb852434830fb92185c3ab49a82d062/update.sh) and you are only scraping the page source of GitHub Package directly without using docker manifests. Is it used in commits that have not yet been pushed in version support? Anyway, this may be broken in the future due to modification of the page structure.

The version and number of downloads can be obtained from the HTML source, but the size must be referenced in the docker manifest.

---

Based on my research for now:

||GHCR Docker Registry HTTP API|GitHub REST API|GitHub GraphQL API|Scraping Package Page|
|-|:-:|:-:|:-:|:-:|
|Versions|o|[o](https://docs.github.com/en/rest/packages/packages?apiVersion=2022-11-28#list-package-versions-for-a-package-owned-by-a-user)|x|o|
|Image size|o|x|x|x|
|Downloads|x|x|x|o|

**Comment by ipitio** [comment]
I've got size working using the manifest on the dev branch, where I've implemented version support too. Will merge the changes to master soon!

https://github.com/ipitio/ghcr-pulls/blob/dev/update.sh

**Comment by TheCaptain989** [comment]
> @TheCaptain989 I've got version support mostly implemented!

Awesome.
I got tags/versions working for my purposes: https://github.dev/TheCaptain989/ghcr-pulls/

**Comment by douglasparker** [comment]
I guess I’m a little confused why I am getting emails about another project within this issue. Wouldn’t it be better that you open a discussion on your own repo?

**Comment by ipitio** [comment]
Sorry! I've made a discussion here: https://github.com/ipitio/ghcr-pulls/discussions/9

**Comment by kaysond** [comment]
Bump. would love to have this!

**Comment by DCjanus** [comment]
I spent some time looking into one possible `version` meaning: resolve `latest`, then show the version tag for the same image.

For example:

```text
sha256:aaa -> latest, v2.1.0
sha256:bbb -> v1.9.0
sha256:ccc -> v1.8.0
```

In this case I would expect `v2.1.0`.

The missing piece is the reverse mapping. The registry API gives us `tags/list` and manifests, but I do not see a `digest -> tags` lookup. So this would require listing candidate tags and checking manifest digests one by one.

I also considered a simpler fallback: ignore `latest` and return the largest semver-like tag. That avoids digest matching, but `tags/list` does not support sorting or server-side filtering, and the result is paginated.

So both registry-only approaches need clear limits. Reading one page can be wrong; reading everything can be too expensive for a public badge.

**Comment by felipecrs** [comment]
I would still rely on parsing the labels:

```console
❯ regctl image inspect ghcr.io/renovatebot/renovate:latest --format '{{if index .Config.Labels "org.label-schema.version"}}{{index .Config.Labels "org.label-schema.version"}}{{else if index .Config.Labels "org.opencontainers.image.version"}}{{index .Config.Labels "org.opencontainers.image.version"}}{{end}}'
43.204.1
```

**Comment by reneleonhardt** [comment]
> I would still rely on parsing the labels:
> 
> ❯ regctl image inspect ghcr.io/renovatebot/renovate:latest --format '{{if index .Config.Labels "org.label-schema.version"}}{{index .Config.Labels "org.label-schema.version"}}{{else if index .Config.Labels "org.opencontainers.image.version"}}{{index .Config.Labels "org.opencontainers.image.version"}}{{end}}'
> 43.204.1

If you don't want to call a binary, https://ghcr.io/renovatebot/renovate:latest redirects to https://github.com/renovatebot/renovate/pkgs/container/renovate, which returns HTML with the labels `latest` `43.205.2` `43.205` `43` for the respective digest.

**Comment by felipecrs** [comment]
It should be possible to use registry APIs to get the same data, though.

**Comment by kaysond** [comment]
Threw this up hoping we can an endpoint from github to make this trivial: https://github.com/orgs/community/discussions/194174
